framework/src/outputs/Checkpoint.C

```diff
@@ -72,6 +80,67 @@ Checkpoint::directory() const
   return _file_base + "_" + _suffix;
 }
 
+void
+Checkpoint::outputStep(const ExecFlagType & type)
+{
+  // Output is not allowed
+  if (!_allow_output && type != EXEC_FORCED)
+    return;
+
+  // If recovering disable output of initial condition, it was already output
+  if (type == EXEC_INITIAL && _app.isRecovering())
+    return;
+
+  // Check whether we should output, then do it.
+  if (shouldOutput(type))
+  {
+    TIME_SECTION("outputStep", 2, "Outputting Step");
+    output(type);
+  }
+}
+
+bool
+Checkpoint::shouldOutput(const ExecFlagType & type)
+{
+  switch (AUTOSAVE_TYPE(_is_autosave))
+  {
+    case NONE:
+      // Only call for onInterval now if it isn't an autosave.
+      if (onInterval() || type == EXEC_FINAL)
+        return Output::shouldOutput(type);
+      return false;
+
+    case SYSTEM_AUTOSAVE:
+      // if this is a pure system-created autosave through AutoCheckpointAction,
+      // then sync across processes and only output one time per signal received.
+      comm().max(Moose::autosave_flag);
+      if (Moose::autosave_flag)
+      {
+        Moose::autosave_flag = 0;
+        return true;
+      }
+      return false;
+
+    case MODIFIED_EXISTING:
+      // First, check if it is a normal time the checkpoint should output.
+      // If it is, we're done here, and we do the normal checks.
+      if (onInterval() || type == EXEC_FINAL)
+        return Output::shouldOutput(type);
+
+      // only run and check comm().max if it isn't a normal output,
+      // i.e. we are running on a non-interval timestep
+      comm().max(Moose::autosave_flag);
+      if (Moose::autosave_flag)
+      {
+        Moose::autosave_flag = 0;
+        return true;
+      }
```

Comment:
I think you've worn me down enough on this PR that I won't push back on this. However, I would like to follow up on another PR and rework `shouldOutput` and `outputStep`. The whole thing is a bit messy, which is why you have a mix of copied code and more code paths than necessary now. I think this could all be cleaned up a lot!

Reply:
Yeah, I want to focus on just getting the signal handling in before we worry about making it all clean. It's obviously ugly but don't want to tackle too many things in one PR.

Reply:
Managed to make this prettier and avoid duplication with a lambda helper. Check out how it is now.